### PR TITLE
feature(zms): add new league middlewares and other models and tools. 

### PR DESCRIPTION
**Pull request by hiha3456**
## Description
1. add LeagueCoordinator, LeagueLearnerCommunicator, StepLeagueActor, BattleStepCollector, battle_inferencer, battle_rolloutor, with their corresponding tests.
2. add BattleTransitionList to gather transitions and cut trajectories for league environment
3. add dataclass of actor data and learner model
4. add an attribute return_original_data into EnvSupervisor
5. add BattleContext into context.py
6. add EventEnum and add feature in event_loop so that we can add customized string in chosen events.
7. add utils sparse_logging to logging in a sparse frequency
8. add my_pickle_loads inside ding/framework/parallel.py so that we could transfer a cuda tensor to a pure-cpu node without bug
9. add old Storage and FileStorage class used in old dev-league branch inside ding/framework/storage
10. change a bit player
11. add sl_branch inside starcraft_player
12. add old ding/league/v2/base_league.py used in old dev-league branch
13. add steve, change upgo, vtrace in ding/rl_utils
14. add detach_grad, flatten in ding/torch_utils/data_helper.py
15. add l2_distance in ding/torch_utils/metric.py
16. add GLU2, GatedConvResBlock, scatter_connection_v2, AttentionPool, lstm ding/torch_utils/network/, and make some changes in networks
17. add read_yaml_config in ding/utils/default_helper.py
18. and other changes...

## TODO
1. delete old checkpoints saved by LeagueLearnerCommunicator, because it can consume disk storage very quickly
2. When we use multiple envs inside env_manager(or other variants), in some cases only some of the envs work properly, so for the return timesteps of step(action), I hugely recommend that, **all the EnvManagers need to return the timesteps in format dict instead of list**. As far as I know, the return of EnvSupervisor and BaseEnvManagerV2 is dict, and the returns subprocesEnvManager and BaseEnvManager is list.
3. The BattleCollector now could not handle the case when the policy has **intermediate state**, for example, the policy of SC2(which in DI-star) maintain a huge amount of intermediate state. In this case, **each env should maintain number of players policies**, which is not the case in current BattleCollector. The current BattleCollector can only handle this kind of policy when EnvManager has only one environment.
4. add middleware of teacher model.

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/rl_utils/steve.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+from typing import List, Callable
```
Comment (PaParaZz1) [inline ding/rl_utils/steve.py:1]: why add this file

File: ding/torch_utils/data_helper.py [hunk 2]
```diff
@@ -397,3 +397,32 @@ def get_null_data(template: Any, num: int) -> List[Any]:
         data['reward'].zero_()
         ret.append(data)
     return ret
+
```
Comment (PaParaZz1) [inline ding/torch_utils/data_helper.py:400]: add unittesr for these two functions

File: ding/torch_utils/metric.py [hunk 3]
```diff
@@ -2,6 +2,16 @@
 from typing import Optional, Callable
 
 
+def l2_distance(a, b, min=0, max=0.8, threshold=5, spatial_x=160):
```
Comment (PaParaZz1) [inline ding/torch_utils/metric.py:5]: add python typing lint and unittest for this function

File: ding/torch_utils/network/activation.py [hunk 4]
```diff
@@ -91,3 +92,31 @@ def build_activation(activation: str, inplace: bool = None) -> nn.Module:
         return act_func[activation]
     else:
         raise KeyError("invalid key for activation: {}".format(activation))
+
+
+class GLU2(nn.Module):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/activation.py:97]: add comments about the difference between `GLU2` and `GLU`

File: ding/utils/data/collate_fn.py [hunk 5]
```diff
@@ -80,10 +84,10 @@ def default_collate(batch: Sequence,
             out = elem.new(storage)
         if elem.shape == (1, ) and cat_1dim:
             # reshape (B, 1) -> (B)
-            return torch.cat(batch, 0, out=out)
+            return torch.cat(batch, dim, out=out)
             # return torch.stack(batch, 0, out=out)
         else:
-            return torch.stack(batch, 0, out=out)
+            return torch.stack(batch, dim, out=out)
```
Comment (PaParaZz1) [inline ding/utils/data/collate_fn.py:90]: add unittest for this argument

File: ding/envs/env_manager/env_supervisor.py [hunk 6]
```diff
@@ -255,6 +260,8 @@ def ready_obs(self) -> tnp.array:
             >>> timesteps = env_manager.step(action)
         """
         active_env = [i for i, s in self._env_states.items() if s == EnvState.RUN]
+        if self._return_original_data:
```
Comment (PaParaZz1) [inline ding/envs/env_manager/env_supervisor.py:263]: why add this

File: ding/framework/context.py [hunk 7]
```diff
@@ -75,3 +75,38 @@ def __init__(self, *args, **kwargs) -> None:
         self.last_eval_iter = -1
 
         self.keep('train_iter', 'last_eval_iter')
+
+
+class BattleContext(Context):
```
Comment (PaParaZz1) [inline ding/framework/context.py:80]: add overview comments

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/envs/env_manager/env_supervisor.py (hunk 6 above)
Comment (hiha3456) [reply ding/envs/env_manager/env_supervisor.py:263]: In the comment, I explained it as "return original observation,  so that the attribute self._ready_obs is not a tnp.array but only the original observation, and the property self.ready_obs is a dict in which the key is the env_id." In fact the reason why I add this attribute is:
1. Sometimes for multi-envs we cannot get observation for every envs, so a dict instead of list is necessary, to directly get the env_id of obs. **PS**: And it is strange that,  for the return timesteps of step(action), the return of EnvSupervisor and BaseEnvManagerV2 is list, and the returns subprocesEnvManager and BaseEnvManager is dict. I think we need to unify them to dict.
2. For envs like DI-star, the tnp.array will raise bug in EnvSupervisor, I forgot which specific bug it is, but it made me to return directly obs instead of tnp.array.  I will try to reproduce this bug if necessary

File: ding/utils/tests/test_sparse_logging.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import pytest
```
Comment (PaParaZz1) [inline ding/utils/tests/test_sparse_logging.py:1]: use pytest

File: ding/utils/sparse_logging.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+from ditk import logging
+import itertools
+import timeit
+
+_log_counter_per_token = {}
+_log_timer_per_token = {}
+
+
+def _get_next_log_count_per_token(token):
+    """Wrapper for _log_counter_per_token. Thread-safe.
```
Comment (PaParaZz1) [inline ding/utils/sparse_logging.py:10]: modify comment format

File: ding/torch_utils/network/scatter_connection.py [hunk 10]
```diff
@@ -91,3 +93,29 @@ def forward(self, x: torch.Tensor, spatial_size: Tuple[int, int], location: torc
         output = output.reshape(N, B, H, W)
         output = output.permute(1, 0, 2, 3).contiguous()
         return output
+
+
+def scatter_connection_v2(shape, project_embeddings, entity_location, scatter_dim, scatter_type='add'):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/scatter_connection.py:98]: can we merge v2 to original ScatterConnection

File: ding/torch_utils/network/activation.py [hunk 11]
```diff
@@ -91,3 +92,31 @@ def build_activation(activation: str, inplace: bool = None) -> nn.Module:
...
+
+    def forward(self, x, context):
+        gate = self.layer1(context)
+        gate = torch.sigmoid(gate)
+        x = gate * x
+        x = self.layer2(x)
+        return x
+
+
+def build_activation2(activation):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/activation.py:117]: we should merge this function into original version

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/framework/middleware/league_coordinator.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+from ditk import logging
+
+from ding.utils.sparse_logging import log_every_sec
+
+if TYPE_CHECKING:
+    from easydict import EasyDict
+    from ding.framework import Task, Context
+    from ding.league.v2 import BaseLeague
+    from ding.league.player import PlayerMeta
+    from ding.league.v2.base_league import Job
```
Comment (PaParaZz1) [inline ding/framework/middleware/league_coordinator.py:16]: on import from 2-level directory, like `from ding.league.v2 import BaseLeague, Job`

File: ding/framework/middleware/league_coordinator.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        self.league.update_payoff(job)
+
+    def __del__(self):
+        logging.info("[Coordinator {}] all tasks finished, coordinator closed".format(task.router.node_id))
+
+    def __call__(self, ctx: "Context") -> None:
+        sleep(1)
+        log_every_sec(
+            logging.INFO, 600, "[Coordinator {}] running jobs {}".format(task.router.node_id, self._running_jobs)
+        )
```
Comment (PaParaZz1) [inline ding/framework/middleware/league_coordinator.py:88]: maybe we should add state_dict/load_state_dict method for coordinator?

File: ding/framework/middleware/tests/mock_for_test.py [hunk 14]
```diff
@@ -116,3 +122,211 @@ def __init__(self) -> None:
...
+    def update_active_player(self, meta):
+        self.update_active_player_cnt += 1
+
+    def create_historical_player(self, meta):
+        self.create_historical_player_cnt += 1
+
+    def get_job_info(self, player_id):
+        self.get_job_info_cnt += 1
+        other_players = [i for i in self.active_players_ids if i != player_id]
+        another_palyer = random.choice(other_players)
```
Comment (PaParaZz1) [inline ding/framework/middleware/tests/mock_for_test.py:148]: typo

File: ding/framework/parallel.py [hunk 15]
```diff
@@ -15,11 +16,40 @@
...
+            return super().find_class(module, name)
+
+
+def cpu_loads(x):
+    bs = io.BytesIO(x)
+    unpickler = CpuUnpickler(bs)
+    return unpickler.load()
+
+
+def my_pickle_loads(msg):
```
Comment (PaParaZz1) [inline ding/framework/parallel.py:40]: move this function to `ding/utils`

File: ding/framework/middleware/functional/actor_data.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+@dataclass
+class ActorData:
+    meta: ActorDataMeta
+    train_data: List[ActorEnvTrajectories] = field(default_factory=[])
+
+
+@dataclass
+class PlayerModelInfo:
+    get_new_model_time: float
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/actor_data.py:28]: why two time here, maybe we can simplify them

File: ding/framework/middleware/functional/collector.py [hunk 17]
```diff
@@ -1,13 +1,22 @@
...
 from functools import reduce
 import treetensor.torch as ttorch
-from ding.envs import BaseEnvManager
+from ding.envs.env_manager.base_env_manager import BaseEnvManager
+from ding.envs.env.base_env import BaseEnvTimestep
 from ding.policy import Policy
-from ding.torch_utils import to_ndarray
+import torch
+from ding.utils import dicts_to_lists
+from ding.torch_utils import to_tensor, to_ndarray
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:10]: policy import sequence

File: ding/framework/middleware/functional/collector.py [hunk 18]
```diff
@@ -45,6 +54,125 @@ def clear(self):
...
+        return return_episode  # list of trajectories
+
+    def clear_newest_episode(self, env_id: int, before_append=False) -> None:
+        # Call this method when env.step raise some error
+
+        # If call this method before append, and the last episode of this env is done,
+        # it means that the env had some error at the first step of the newest episode,
+        # and we should not delete the last episode because it is a normal episode.
+        if before_append is True and len(self._done_episode[env_id]) > 0 and self._done_episode[env_id][-1] == True:
+            return 0
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:142]: what is the usage of return value here

File: ding/framework/middleware/functional/collector.py [hunk 19]
```diff
@@ -45,6 +54,125 @@ def clear(self):
...
+        return all_env_data
+
+    def _cut_trajectory_from_episode(self, episode: list) -> List[List]:
+        # first we cut complete trajectories (list of transitions whose length equal to unroll_len)
+        # then we gather the transitions in the tail of episode,
+        # and fill up the trajectory with the tail transitions in Trajectory(t-1)
+        # If we don't have Trajectory(t-1), i.e. the length of the whole episode is smaller than unroll_len,
+        # we fill up the trajectory with the first element of episode.
+        return_episode = []
+        i = 0
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:110]: why init `i` here

File: ding/framework/middleware/functional/collector.py [hunk 20]
```diff
@@ -45,6 +54,125 @@ def clear(self):
...
+
+        if num_tail_transitions > 0:
+            trajectory = episode[-self._unroll_len:]
+            if len(trajectory) < self._unroll_len:
+                initial_elements = []
+                for _ in range(self._unroll_len - len(trajectory)):
+                    initial_elements.append(trajectory[0])
+                trajectory = initial_elements + trajectory
+            if self._last_step_fn:
+                last_step = deepcopy(trajectory[-1])
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:128]: why deepcopy here

File: ding/framework/middleware/league_actor.py [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+        collector = self._get_collector(job.launch_player)
+
+        _, ctx.current_policies = self._get_current_policies(job)
+
+        ctx.n_episode = self.cfg.policy.collect.n_episode
+        assert ctx.n_episode >= self.env_num, "[Actor {}] Please make sure n_episode >= env_num".format(
+            task.router.node_id
+        )
+
+        ctx.n_episode = self.cfg.policy.collect.n_episode
```
Comment (PaParaZz1) [inline ding/framework/middleware/league_actor.py:166]: repeat?

File: ding/framework/middleware/league_actor.py [hunk 22]
```diff
@@ -0,0 +1,228 @@
...
+        assert ctx.n_episode >= self.env_num, "Please make sure n_episode >= env_num"
+
+        ctx.episode_info = [[] for _ in range(self.agent_num)]
+
+        while True:
+            time_begin = time.time()
+            collector(ctx)
+
+            if ctx.job_finish is True:
```
Comment (PaParaZz1) [inline ding/framework/middleware/league_actor.py:175]: if ctx.job_finish

File: ding/framework/middleware/league_actor.py [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+            log_every_sec(
+                logging.INFO, 5,
+                '[Actor {}] sent {} trajectories till now, total trajectory send speed is {} traj/s'.format(
+                    task.router.node_id,
+                    self.traj_num,
+                    self.traj_num / self.total_time,
+                )
+            )
+
+            gc.collect()
```
Comment (PaParaZz1) [inline ding/framework/middleware/league_actor.py:203]: why call gc here

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]
